### PR TITLE
enh: avoid using duplicate variable in getter function when in same module

### LINKED ISSUE #4125: bug: when setting array dimension through a program variable, modifying the variable does not change the dimension
I think this is what @assem2002 had wanted to point in https://github.com/lfortran/lfortran/pull/4104#issuecomment-2142798506. This is loosely related to #2430 too.

The issue occurs only in the `PROGRAM` scope when array arguments to a subroutine are set using a variable. It is observed that modifying the variable does not modify the dimension.

```fortran
program example
   implicit none
   integer(4), dimension(6) :: cs
   integer(4), save :: nx = 4

   cs = [1, 2, 3, 4, 5, 6]
   call f(cs)
   nx = 10
   call f(cs)
   
   contains
   
      subroutine f(x)
         integer(4), dimension(nx), intent(in) :: x
         print *, x
      end subroutine f

end program example
```
```console
(base) saurabh-kumar@Awadh:~/Projects/System/lfortran$ lfortran ./examples/example.f90
1 2 3 4 
1 2 3 4 
```

The chief cause of the issue is that the transformed code hardcodes the value of the variable `nx` to store `4` and so we always fetch that value.

Here is the transformed code:
```fortran
program example
   implicit none
   integer(4), dimension(6) :: cs
   integer(4), save :: nx = 4
   cs = [1, 2, 3, 4, 5, 6]
   call f(cs)
   nx = 10
   call f(cs)

contains

   subroutine f(x)
      integer(4), dimension(__lcompilers_get_nx()), intent(in) :: x
      interface
         pure integer(4) function __lcompilers_get_nx()
         end function __lcompilers_get_nx

      end interface
      print *, x
   end subroutine f

   pure integer(4) function __lcompilers_get_nx()
   integer(4), save :: nx = 4        ! the value 4 is hardcoded here
   __lcompilers_get_nx = nx
   end function __lcompilers_get_nx

end program example
```

The issue does not occur when we use a module variable for setting the dimension. Please see https://github.com/lfortran/lfortran/pull/4104#issuecomment-2142857084.

### LINKED ISSUE #4498: Bug : getter functions do hard code assignment
**MRE**

```Fortran
module A
    IMPLICIT NONE
    integer :: i = 4
    integer :: arr_mod(4,4)
    contains
    subroutine sub (arr)
        integer , INTENT(OUT):: arr(4,i)
        print *,size(arr)            ! This prints 16 while it should print 20000
    end subroutine sub
end module A


program mainnn
    use A
    i = 5000
    call sub(arr_mod)
end program
```

The getter function gets created with the variable hard coded to the initial value at compile time, then if the value gets changed in the compile time the function becomes usless.
normally getter function works normally when it tries to get an external variable, but when it tries to access a variable that's isn't an external one the described issue occurs.

### PR DESCRIPTION
**Pull request by assem2002**
fixes #4498  
closes #4125 
Thanks @kmr-srbh for the test.
***
We avoid duplicating the variable and adding it into `lcompilers_getter_...()` when it isn't an `ExternalSymbol` which means it's a symbol in the same scope and doesn't need that kind of processing.
***
**concerns**
- I'm assuming the passed `end_sym` is not `ExternalSymbol` when it's a symbol that's already declared in the same module, program or function. If that assumption is wrong in some cases, this fix won't be efficient.

### CODE REVIEW

**Comment by kmr-srbh** [comment]
Could you please post the transformed code output for the MRE shared in #4125?

**Comment by kmr-srbh** [comment]
Also, let's close either of #4498 or #4125 to simplify things as they are the same.

**Comment by assem2002** [comment]
for the MRE in #4125 HERE is the fortran after creating ASR :
```FORTRAN
! Fortran code after Body Visitor
program example
implicit none
integer(4), dimension(6) :: cs
integer(4), save :: nx = 4
cs = [1, 2, 3, 4, 5, 6]
call f(cs)
nx = 10
call f(cs)

contains

pure integer(4) function __lcompilers_get_nx()
    __lcompilers_get_nx = nx
end function __lcompilers_get_nx

subroutine f(x)
    integer(4), dimension(__lcompilers_get_nx()), intent(in) :: x
    print *, x
end subroutine f

end program example


```

**Comment by kmr-srbh** [comment]
The changes in this PR looks good to me! :+1:

File: src/lfortran/semantics/ast_common_visitor.h [hunk 1]
```diff
@@ -1206,11 +1206,11 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
-
-        // populate symbol table
-        ASRUtils::SymbolDuplicator sd(al);
-        sd.duplicate_symbol(end_sym, current_scope);
-        end_sym = current_scope->resolve_symbol(ASRUtils::symbol_name(end_sym));
+        if (ASR::is_a<ASR::ExternalSymbol_t>(*end_sym)) {
+            ASRUtils::SymbolDuplicator sd(al);
+            sd.duplicate_symbol(end_sym, current_scope);
+            end_sym = current_scope->resolve_symbol(ASRUtils::symbol_name(end_sym));
+        }
```
Comment (Pranavchiku) [inline src/lfortran/semantics/ast_common_visitor.h:1213]: Yeah, this is correct. It took me time to figure out what you are trying to do.

**Comment by Pranavchiku** [comment]
> I'm assuming the passed end_sym is not ExternalSymbol when it's a symbol that's already declared in the same module, program or function. If that assumption is wrong in some cases, this fix won't be efficient.


Here you go,

```fortran
module A
    IMPLICIT NONE
    integer :: i = 4
    integer :: arr_mod(4,4)
    contains
    subroutine sub (arr)
        integer , INTENT(OUT):: arr(4,i)
        print *,size(arr)            ! This prints 16 while it should print 20000
    end subroutine sub
end module A

subroutine sub2 (arr)
   use A
   integer , INTENT(OUT):: arr(4,i)
   print *,size(arr)            ! This prints 16 while it should print 20000
end subroutine sub2


program mainnn
    use A
    i = 5000
    call sub(arr_mod)
    call sub2(arr_mod)
end program
```

Your fix will work, because it will now create an `ExternaSymbol` and not the node equivalent to `integer, save :: i = 4` as shown below

```clojure
                            i:
                                (Variable
                                    2
                                    i
                                    []
                                    Local
                                    (IntegerConstant 4 (Integer 4) Decimal)
                                    (IntegerConstant 4 (Integer 4) Decimal)
                                    Save
                                    (Integer 4)
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                ),
````

**Review by Pranavchiku — APPROVED** [review]
Thank you for fixing this! This was a great catch.

**Pull request merged**